### PR TITLE
Experiment with representing dimension values in the type system

### PR DESCRIPTION
**Pull request by honnibal**
We can't expect the type checking to provide full validation over literal types, as we won't be able to express something like "concatenate returns output with dimensionality equal to the sum of its layers' dimensionality. However, even if the checking doesn't really work properly, it's worth discussing what this could look like, and what we can and can't do with it.

I'm not 100% sure I'm doing this right, but I've tried to make the `Floats1d`, `Floats2d` etc arrays into generics, giving us space to pass type variables in to represent the dimensions. I've taken care to use distinct variables for each class. I'm not sure this is actually necessary? I didn't want mypy to think the dimensions across the classes matched the same values or something.

Arguably we could avoid having the `Floats1d`, `Floats2d` etc classes, and make these generic with respect to number of dimensions. I think that's probably worse: I think it's worth being able to express `Floats2d` concisely, and to me it seems to make things simpler and easier to work with. If it turns out making this a parametric type with a literal is better, sure we can do that.

The main thing that I think is neat is having a way to document the expected dimensionalities in the code.

I have no idea what conventions we should use to distinguish the dimension type variables from real variables, though. We want the type variables to have concise names, but we still want concise names available within the code. I think we need a better variable convention in general around this stuff. For instance, maybe weights parameters should be named like `wW`, `wB` by convention? Then their gradients could still be `dW`.

### CODE REVIEW

**Comment by honnibal** [comment]
Hm, Python 3.8 fails due to method resolution order? Seems hard to fix =/

File: thinc/layers/relu.py [hunk 1]
```diff
@@ -17,7 +21,7 @@ def ReLu(
     dropout: Optional[float] = None,
     normalize: bool = False,
 ) -> Model:
-    model = Model[Floats2d, Floats2d](
+    model = Model[Floats2d[B, O], Floats2d[B, O]](
```
Comment (svlandeg) [inline thinc/layers/relu.py]: should the first one be `[B, I]` instead of `[B, O]` ?

**Comment by svlandeg** [comment]
It seems to work when you put `Array` first like you did for `Floats1d` and `Floats2d` ?

**Comment by honnibal** [comment]
Closing this: I think the experiment really leads to the conclusion that the dimension-based typing isn't a good approach, certainly for now, and possibly in general. I still find the documentation aspect pretty appealing, but the complexity just doesn't work, and it messes up the type variables we want to use for too many methods.

**Pull request closed without merging**